crates/net/network/tests/it/connect.rs

```diff
@@ -80,3 +86,50 @@ async fn test_connect_with_boot_nodes() {
         dbg!(ev);
     }
 }
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_incoming_node_id_blacklist() {
+    reth_tracing::init_tracing();
+    let secret_key = SecretKey::new(&mut rand::thread_rng());
+
+    let reth_socket = SocketAddr::new([127, 0, 0, 1].into(), 30305);
+
+    // instantiate geth and add ourselves as a peer
+    let temp_dir = tempfile::tempdir().unwrap().into_path();
+    let geth = Geth::new().data_dir(temp_dir).disable_discovery().spawn();
+    let geth_endpoint = SocketAddr::new([127, 0, 0, 1].into(), geth.port());
+    let provider = Provider::<Http>::try_from(format!("http://{geth_endpoint}")).unwrap();
+
+    // get the peer id we should be expecting
+    let geth_peer_id: PeerId =
+        provider.node_info().await.unwrap().enr.public_key().encode_uncompressed().into();
+
+    let ban_list = BanList::new(HashSet::from_iter(vec![geth_peer_id]), HashSet::default());
+    let peer_config = PeersConfig::default().with_ban_list(ban_list);
+
+    let config = NetworkConfig::builder(Arc::new(TestApi::default()), secret_key)
+        .listener_addr(reth_socket)
+        .peer_config(peer_config)
+        .build();
+
+    let network = NetworkManager::new(config).await.unwrap();
+
+    let handle = network.handle().clone();
+    tokio::task::spawn(network);
+
+    // make geth connect to us
+    let our_peer_id = handle.peer_id();
+    let our_enode = format!("enode://{}@{}", hex::encode(our_peer_id.0), reth_socket);
+
+    provider.add_peer(our_enode).await.unwrap();
+
+    let events = handle.event_listener();
+    let mut event_stream = NetworkEventStream::new(events);
+
+    // check for session to be opened
```

Comment:
Should we be doing the banlist check before even the session is opened so we don't need to open and then immediately close? Or is that not possible? @mattsse

Reply:
for the case with incoming banned  `PeerIds` afaik we need to handshake them to see there id. all other cases we close before opening a session

Reply:
thanks - makes sense!